Objective-C/Internal/Replicator/CBLWebSocket.mm

```diff
@@ -372,63 +370,79 @@ - (void) connectToHostWithName: (NSString*)hostname
-    
-    int res = getaddrinfo(cHost, cPort, &hints, &_addr);
-    if (res) {
-        NSString* msg = $sprintf(@"Failed to get address info with error %d", res);
-        CBLWarnError(WebSocket, @"%@: %@", self, msg);
-        [self closeWithError: addrInfoError(res, msg)];
-        return;
-    }
     
-    CBLLogVerbose(WebSocket, @"%@: Host '%@' was resolved as %@", self, hostname, addrInfo(_addr));
-    
-    if (_addr->ai_family != AF_INET && _addr->ai_family != AF_INET6) {
-        NSString* msg = $sprintf(@"Address family %d is not supported", _addr->ai_family);
+    unsigned int index = if_nametoindex([interface cStringUsingEncoding: NSUTF8StringEncoding]);
+    if (index == 0) {
+        int errNo = errno;
+        NSString* msg = $sprintf(@"Failed to find network interface %@ with errno %d", interface, errNo);
         CBLWarnError(WebSocket, @"%@: %@", self, msg);
-        [self closeWithError: posixError(EPERM, msg)];
+        [self closeWithError: posixError(errNo, msg)];
         return;
     }
-    
-    // Create socket:
-    [self _socketConnect: interface];
+
+    _dnsService = [[CBLDNSService alloc] initWithHost: hostname
+                                            interface: index
+                                                 port: (UInt16)port
+                                             delegate: self];
+    [_dnsService start];
 }
 
+#pragma mark DNSServiceDelegate
+
+ - (void) didResolveSuccessWithAddress: (AddressInfo*)info {
+     dispatch_async(_queue, ^{
+         if (_dnsService) {
+             CBLLogVerbose(WebSocket, @"%@: Host '%@' was resolved as ip=%@, family=%d",
+                           self, info.host, info.addrstr, info.addr->sa_family);
+             [self _socketConnect: info];
+         }
+     });
+ }
+
+ - (void) didResolveFailWithError: (NSError*)error {
+     dispatch_async(_queue, ^{
+         if (_dnsService) {
+             CBLWarnError(WebSocket, @"%@: Host was failed to resolve with error '%@'",
+                           self, error.my_compactDescription);
+             [self closeWithError: error];
+         }
+     });
+ }
+
 #pragma mark - Socket connect
 
-- (void) _socketConnect: (NSString*)interface {
+- (void) _socketConnect: (AddressInfo*)info {
     Assert(_sockfd < 0);
-    _sockfd = socket(_addr->ai_family, _addr->ai_socktype, _addr->ai_protocol);
+    _sockfd = socket(info.addr->sa_family, SOCK_STREAM, 0);
     if (_sockfd < 0) {
         int errNo = errno;
-        NSString* msg = $sprintf(@"Failed to create socket with errno %d (%@)", errNo, addrInfo(_addr));
+        NSString* msg = $sprintf(@"Failed to create socket with errno %d (%@)", errNo, info);
         CBLWarnError(WebSocket, @"%@: %@", self, msg);
         [self closeWithError: posixError(errNo, msg)];
         return;
     }
     
     // Set network interface:
-    if (interface) {
-        NSError* err;
-        bool useIPv4 = (_addr->ai_family == AF_INET);
+    if (_networkInterface) {
```

Comment:
There is no case that this will be nil. You could also just do the assert that `info.interface > 0`.